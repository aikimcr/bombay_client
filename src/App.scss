@import './global.scss';

@mixin areaLayout($area, $bgColor, $fgColor, $width: 100%, $height: 100%, $top: 0, $left: 0, $right: null, $bottom: null) {
  grid-area: $area;
  position: relative;
  overflow: hidden;

  @if ($bottom ==null) {
    top: $top;
  }

  @else {
    bottom: $bottom;
  }

  @if ($right ==null) {
    left: $left;
  }

  @else {
    right: $right;
  }

  width: $width;
  height: $height;
  background-color: $bgColor;
  color: $fgColor;
}

.App {
  width: 100%;
  height: 100%;
  min-width: 800px;
  min-height: 500px;
  overflow: hidden;
  display: grid;
  grid-template-columns: 150px auto 50px;
  grid-template-rows: 60px 30px auto 30px;
  grid-template-areas:
    "header header header"
    "navigation navigation navigation"
    "filters content accessories"
    "footer footer footer";

}

.header-bar {
  @include areaLayout(header, $header-background-color, $header-color, $height: 60px);
}

.app-navigation {
  @include areaLayout(navigation, $navigation-background-color, $navigation-color, $height: 30px);
  font-size: 20px;
  line-height: 20px;
}

.filters {
  @include areaLayout(filters, $sidebar-background-color, $sidebar-color, 150px);
}

.content {
  @include areaLayout(content, white, black);
}

.accessories {
  @include areaLayout(accessories, $sidebar-background-color, $sidebar-color, 50px, $right: 0);
}

.footer {
  @include areaLayout(footer, $header-background-color, $header-color, $height: 30px);
}