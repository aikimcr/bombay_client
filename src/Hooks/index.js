export * from './useLoginTracking';
export * from './useModelCollection';
