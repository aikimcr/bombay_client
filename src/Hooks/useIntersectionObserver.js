import { useEffect, useRef } from "react";

export function useIntersectionObserver(topRef, callback, options = {}) {
  const observer = useRef(null);

  useEffect(() => {
    if (topRef.current == null) return;

    if (observer.current == null) {
      observer.current = new IntersectionObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      observer.current.disconnect();
    };
  }, [callback, options, topRef]);

  return observer;
}

export default useIntersectionObserver;
