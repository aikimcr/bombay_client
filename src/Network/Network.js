// Manage XHR request calls.  This is the low-level stuff.
// import serverConfig from './serverConfig';
const serverConfig = {};

const serverConfigKeys = {
  serverProtocol: 'REACT_APP_SERVER_PROTOCOL',
  serverHost: 'REACT_APP_SERVER_HOST',
  serverBasePath: 'REACT_APP_SERVER_BASE_PATH',
  serverPort: 'REACT_APP_SERVER_PORT',
}

for (const key in serverConfigKeys) {
  if (process.env.hasOwnProperty(serverConfigKeys[key])) {
    const value = process.env[serverConfigKeys[key]];

    switch(value) {
      case 'null':
      case 'none':
      case 'empty':
        serverConfig[key] = null;
        break;

      default:
        serverConfig[key] = value;
    }
  }
}

function setConfigOption(key, defaultConfig) {
  return serverConfig.hasOwnProperty(key) ? serverConfig[key] : defaultConfig;
}

export const serverProtocol = setConfigOption('serverProtocol', 'http');
export const serverHost = setConfigOption('serverHost', 'localhost');
export const serverBasePath = setConfigOption('serverBasePath', '');
export const serverPort = setConfigOption('serverPort', 2001);

console.log(`Server Base URL: ${prepareURLFromArgs('')}`);

function getStandardHeaders(includeContentType = true) {
  const result = {};
  if (includeContentType) result['content-type'] = 'application/json';

  const token = localStorage.getItem('jwttoken')

  if (token) {
    result.Authorization = `Bearer ${token}`;
  }

  return result;
}

export function normalizeAndJoinPath(...pathParts) {
  const newPath = pathParts.reduce((memo, part) => {
    if (part === undefined) {
      throw new Error(`Unable to normalize path with parts: "${pathParts.join(', ')}"`);
    }

    let newPart = memo + '/' + part;
    newPart = newPart.replace(/\/\/+/g, '/');

    if (part !== '/') {
      if (newPart.length > 1) {
        newPart = newPart.replace(/\/+$/, '');
      }
    }

    return newPart;
  }, '/');

  return newPath;
}

export function buildURL(args = {}) {
  let { hostname, path, protocol, port } = {
    hostname: serverHost,
    path: '/',
    protocol: serverProtocol,
    port: serverPort,
    ...args,
  }

  hostname = hostname.replace(/^\/+/, '').replace(/\/+$/, '');

  if (hostname.length === 0) {
    hostname = serverHost;
  }

  let result = `${protocol}://${hostname}`;

  if (port) {
    result += `:${port}`;
  }

  if (Array.isArray(path)) {
    path = normalizeAndJoinPath(...path);
  }

  result += path;

  return result;
}

export function prepareURLFromArgs(path, query) {
  const requestUrl = new URL(buildURL());
  requestUrl.pathname = normalizeAndJoinPath(serverBasePath, path);

  for (const param in query) {
    requestUrl.searchParams.set(param, query[param]);
  }

  return requestUrl;
}

async function decodeResponse(response) {
  if (response.ok) {
    const text = await response.text();
    let result = text;

    try {
      result = JSON.parse(text);
    } catch(err) {
      result = text;
    }

    return result;
  }

  return Promise.reject({ status: response.status, message: response.statusText });
}

export async function getFromURLString(urlString) {
  const response = await fetch(urlString, {
    mode: 'cors',
    credentials: 'include',
    headers: getStandardHeaders(false),
  });

  return decodeResponse(response);
}

function buildJSON(body) {
  const sendBody = { ...body };
  delete sendBody.id;
  delete sendBody.url;
  return JSON.stringify(sendBody);
}

export async function postToURLString(urlString, body) {
  const sendJSON = buildJSON(body);
  const response = await fetch(urlString, {
    method: 'POST',
    headers: getStandardHeaders(),
    body: sendJSON,
    mode: 'cors',
    credentials: 'include',
  });

  return decodeResponse(response);
}

export async function putToURLString(urlString, body) {
  const sendJSON = buildJSON(body);
  const response = await fetch(urlString, {
    method: 'PUT',
    headers: getStandardHeaders(),
    body: sendJSON,
    mode: 'cors',
    credentials: 'include',
  });

  return decodeResponse(response);
}