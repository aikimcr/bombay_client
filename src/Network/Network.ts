import { omit } from 'lodash';

// Manage XHR request calls.  This is the low-level stuff.
// import serverConfig from './serverConfig';
export const defaultAPIServer: string = process.env.SERVER;
export const defaultAPIBasePath: string = process.env.SERVER_BASE_PATH;

const getStandardHeaders = (includeContentType = true) => {
  const result: HeadersInit = {};
  if (includeContentType) {
    result['content-type'] = 'application/json';
  }

  const token = localStorage.getItem('jwttoken');

  if (token) {
    result.Authorization = `Bearer ${token}`;
  }

  return result;
};

export const normalizeAndJoinPath = (...pathList: string[]): string => {
  const newPath = pathList.reduce((memo, part) => {
    if (part === undefined) {
      throw new Error(
        `Unable to normalize path with parts: "${pathList.join(', ')}"`,
      );
    }

    let newPart = memo + '/' + part;
    newPart = newPart.replace(/\/\/+/g, '/');

    if (part !== '/') {
      if (newPart.length > 1) {
        newPart = newPart.replace(/\/+$/, '');
      }
    }

    return newPart;
  }, '/');

  return newPath;
};

interface buildUrlProps {
  server?: string;
  basePath?: string;
  applicationPaths?: string[];
}

export const buildURL = ({
  server = defaultAPIServer,
  basePath = defaultAPIBasePath,
  applicationPaths = [],
}: buildUrlProps = {}) => {
  const allPaths = [...applicationPaths];

  if (!server || server === 'default') {
    server = 'http://localhost:2001';
  }

  if (basePath && basePath !== 'default') {
    allPaths.unshift(basePath);
  }

  const result: URL = new URL(server);
  result.pathname = normalizeAndJoinPath(...allPaths);
  return result;
};

export type NetworkURLQueryType = Record<string, string | number | symbol>;

export const prepareURLFromArgs = (
  pathList: string[] = [],
  query: NetworkURLQueryType = {},
) => {
  const requestUrl = buildURL({ applicationPaths: pathList });

  for (const param in query) {
    requestUrl.searchParams.set(param, query[param].toString());
  }

  return requestUrl;
};

const decodeResponse = async (
  response: Awaited<ReturnType<typeof fetch>>,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
): Promise<Record<string, any>> => {
  if (response.ok) {
    const text = await response.text();
    let result = { body: text };

    try {
      result = JSON.parse(text);
    } catch (err) {
      result = { body: text };
    }

    return result;
  }

  return Promise.reject({
    status: response.status,
    message: response.statusText,
  });
};

export const getFromURLString = async (
  urlString: string,
): ReturnType<typeof decodeResponse> => {
  const response = await fetch(urlString, {
    mode: 'cors',
    credentials: 'include',
    headers: getStandardHeaders(false),
  });

  return decodeResponse(response);
};

interface buildJSONProps {
  id?: string | number;
  url?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: any;
}

const buildJSON = (body: buildJSONProps): string => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const sendBody: Record<string, any> = {
    ...omit(body, ['id', 'url']),
  };
  return JSON.stringify(sendBody);
};

export const postToURLString = async (
  urlString: string,
  body: buildJSONProps,
) => {
  const sendJSON = buildJSON(body);
  const response = await fetch(urlString, {
    method: 'POST',
    headers: getStandardHeaders(),
    body: sendJSON,
    mode: 'cors',
    credentials: 'include',
  });

  return decodeResponse(response);
};

export const putToURLString = async (
  urlString: string,
  body: buildJSONProps,
) => {
  const sendJSON = buildJSON(body);
  const response = await fetch(urlString, {
    method: 'PUT',
    headers: getStandardHeaders(),
    body: sendJSON,
    mode: 'cors',
    credentials: 'include',
  });

  return decodeResponse(response);
};
