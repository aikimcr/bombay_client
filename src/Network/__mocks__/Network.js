const RealNetwork = jest.requireActual('../Network');

export const serverProtocol = 'https';
export const serverHost = 'fake';
export const serverBasePath = 'xyzzy';
export const serverPort = null;

export const normalizeAndJoinPath = RealNetwork.normalizeAndJoinPath;

export function buildURL (args = {}) {
    let { hostname, path, protocol, port } = {
        hostname: serverHost,
        path: '/',
        protocol: serverProtocol,
        port: serverPort,
        ...args,
    }

    return RealNetwork.buildURL(args);
}

// ToDo Find a way to do this without having to copy the entire real implementation.
export function prepareURLFromArgs(path, query) {
    const requestUrl = new URL(buildURL());
    requestUrl.pathname = normalizeAndJoinPath(serverBasePath, path);

    for (const param in query) {
        requestUrl.searchParams.set(param, query[param]);
    }

    return requestUrl;
}


export function getFromURLString(url, query) {
    throw new Error(`getFromURLString has not been mocked properly ${url} ${JSON.stringify(query)}`);
}

export async function postToURLString(urlString, body) {
    throw new Error(`postToURLString has not been mocked properly ${url} ${JSON.stringify(body)}`);
}

export async function putToURLString(urlString, body) {
    throw new Error(`putToURLString has not been mocked properly ${url} ${JSON.stringify(body)}`);
}

// Test Utilities not found in the original
let mockPromise;
let mockResolve;
let mockReject;

export function _setupMockPromise() {
    [mockPromise, mockResolve, mockReject] = makeResolvablePromise();
    return {
        promise: mockPromise,
        resolve: mockResolve,
        reject: mockReject,
    };
}

export function _setupMocks() {
    _setupMockPromise();

    this.getFromURLString = jest.fn((urlString, query) => {
        expect(mockPromise).toBeDefined();
        return mockPromise;
    });

    this.postToURLString = jest.fn((urlString, body) => {
        expect(mockPromise).toBeDefined();
        return mockPromise;
    });

    this.putToURLString = jest.fn((urlString, body) => {
        expect(mockPromise).toBeDefined();
        return mockPromise;
    });

    return {
        resolve: mockResolve,
        reject: mockReject,
        promise: mockPromise,
    };
}