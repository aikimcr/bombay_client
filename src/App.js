import { useState, useEffect } from 'react';

import './App.scss';

import BombayContext from './BombayContext';

import HeaderBar from './AppLayout/HeaderBar';
import Navigation from './AppLayout/Navigation';
import Filters from './AppLayout/Filters';
import Content from './AppLayout/Content';
import Accessories from './AppLayout/Accessories';
import Footer from './AppLayout/Footer';

import { loginStatus } from './Network/Login';

function App() {
  const [appState, setAppState] = useState({
    title: 'Bombay Band Management System',
    loggedIn: false,
    mode: 'artist',
  });

  useEffect(() => {
    // ToDo: Make the interval configurable, don't check if logged out.
    const intervalHandle = setInterval(async () => {
      const result = await loginStatus();
      const newState = {
        ...appState,
        loggedIn: result.loggedIn,
      }
      setAppState(newState);
    }, 10 * 60 * 1000); // Ten minutes.

    return () => {
      clearInterval(intervalHandle);
    };
  }, [appState]);

  return (
    <div className="App">
      <BombayContext.Provider value={appState}>
        <HeaderBar />
        <Navigation />
        <Filters />
        <Content />
        <Accessories />
        <Footer />
      </BombayContext.Provider>
    </div >
  );
}

export default App;
