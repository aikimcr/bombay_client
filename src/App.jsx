import React, { useState, useEffect, useCallback } from 'react';
import { BrowserRouter } from 'react-router';

import BombayLoginContext from './Context/BombayLoginContext';
import BombayUtilityContext from './Context/BombayUtilityContext';
import ConfigurationContext from './Context/ConfiguratonContext';

import { fetchBootstrap } from './Network/Bootstrap';
import { useLoginTracking } from './Hooks/useLoginTracking';

import './App.scss';
import {
  AppLayout,
  AppRoutes,
  HeaderLayout,
  LoginStatusDisplay,
  NavigationDropdown,
} from './Components';

export const App = () => {
  const [bootstrap, setBootstrap] = useState(null);
  const [showLoginForm, setShowLoginForm] = useState(false);

  const [loginState, setLoginState] = useLoginTracking();

  const callFetchBootstrap = useEffect(() => {
    (async () => {
      try {
        const result = await fetchBootstrap();
        setBootstrap(result);
        return result;
      } catch (err) {
        console.error(err);
      }
    })();
  }, []);

  const getBootstrap = () => bootstrap;

  const loginContext = {
    loggedIn: loginState,
    showLoginForm: showLoginForm,
    setLoggedIn: (newLoggedIn) => {
      if (newLoggedIn) {
        setShowLoginForm(false);
      }

      setLoginState(newLoggedIn);
    },

    setShowLogin: (newShow) => {
      if (loginState) return;
      setShowLoginForm(newShow);
    },
  };

  const utilities = {
    getBootstrap,
  };

  const routerBase = process.env.CLIENT_ROUTER_BASE || '/';

  const appConfig = {
    routerBase,
  };

  if (!bootstrap) {
    return null;
  }

  return (
    <div className="App" data-testid="app">
      <ConfigurationContext.Provider value={appConfig}>
        <BombayLoginContext.Provider value={loginContext}>
          <BombayUtilityContext.Provider value={utilities}>
            <BrowserRouter basename={routerBase}>
              <AppLayout
                header={
                  <HeaderLayout
                    title="PDQ Band Manager"
                    footer={<LoginStatusDisplay />}
                  />
                }
              >
                <AppRoutes />
              </AppLayout>
            </BrowserRouter>
          </BombayUtilityContext.Provider>
        </BombayLoginContext.Provider>
      </ConfigurationContext.Provider>
    </div>
  );
};
