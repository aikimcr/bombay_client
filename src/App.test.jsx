import { useContext } from "react";
import { act, render, screen } from "@testing-library/react";

import { mockUseLoginTracking } from "./Hooks/testing";

jest.mock("./Hooks/useLoginTracking", () => {
  const originalModule = jest.requireActual("./Hooks/useLoginTracking");

  return {
    __esModule: true,
    ...originalModule,
    useLoginTracking: mockUseLoginTracking,
  };
});

import { mockFetchBootstrap } from "./Network/testing";

jest.mock("./Network/Bootstrap", () => {
  const originalModule = jest.requireActual("./Network/Bootstrap");

  return {
    __esModule: true,
    ...originalModule,
    fetchBootstrap: mockFetchBootstrap,
  };
});

jest.mock("./Components", () => {
  const originalModule = jest.requireActual("./Components");

  return {
    __esModule: true,
    ...originalModule,
    LoginStatusDisplay: () => <div data-testid="mock-login-status">Login</div>,
  };
});

import { App } from "./App";

const testBootstrap = {
  key_signatures: ["A", "B", "C", "D", "E", "F", "G"],
};

describe("Test App", () => {
  test("Compoment matches snapshopt", async () => {
    const bootstrapPromise = PromiseWithResolvers();
    mockFetchBootstrap.mockReturnValue(bootstrapPromise.promise);

    const { asFragment } = render(<App />);

    await act(async () => {
      bootstrapPromise.resolve(testBootstrap);
    });

    expect(asFragment()).toMatchSnapshot();
  });

  test("Renders App Framework", async () => {
    const bootstrapPromise = PromiseWithResolvers();
    mockFetchBootstrap.mockReturnValue(bootstrapPromise.promise);

    await act(async () => {
      render(<App />);
    });

    await act(async () => {
      bootstrapPromise.resolve(testBootstrap);
    });

    expect(screen.getByTestId("app")).toBeInTheDocument();
    expect(screen.getByTestId("mock-login-status")).toBeInTheDocument();
  });
});
