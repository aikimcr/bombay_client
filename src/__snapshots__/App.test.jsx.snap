// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Test App Compoment matches snapshopt 1`] = `
<DocumentFragment>
  <div
    class="App"
    data-testid="app"
  >
    <div
      class="app-layout"
    >
      <header
        class="app-layout__header"
      >
        <div
          class="header-layout"
        >
          <header
            class="header-layout__title"
          >
            Bombay Band Manager
          </header>
          <div
            class="header-layout__content"
          />
          <footer
            class="header-layout__footer"
          >
            <div
              data-testid="mock-login-status"
            >
              Login
            </div>
          </footer>
        </div>
      </header>
      <div
        class="app-layout__content"
      >
        <div
          class="top"
        >
          <h1
            class="top__title"
          >
            Welcome to the Bombay Band Manager System
          </h1>
          <div
            class="top__content"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
